File: 222. Count Complete Tree Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if (root == NULL) return 0;
        int l = lheight(root) + 1;
        int r = rheight(root) + 1;
        if (l == r) return pow(2, l) -1;
        else return countNodes(root->left) + countNodes(root->right) + 1;
    }
    int lheight(TreeNode* root)
    {
        int count = 0;
        while (root->left != NULL)
        {
            root = root->left;
            count ++;
        }
        return count;
    }
    int rheight(TreeNode* root)
    {
        int count = 0;
        while (root->right != NULL)
        {
            root = root->right;
            count ++;
        }
        return count;
    }
};

File: 222. Count Complete Tree Nodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        l = self.lheight(root) + 1
        r = self.rheight(root) + 1
        if l == r:
            return 2 ** l - 1
        return self.countNodes(root.left) + self.countNodes(root.right) + 1
        
    def lheight(self, root):
        count = 0
        while root.left:
            root = root.left
            count += 1
        return count
    
    def rheight(self, root):
        count = 0
        while root.right:
            root = root.right
            count += 1
        return count
